### PR TITLE
Configurable clipboard autoclear timeout support

### PR DESCRIPTION
**Pull request by secrecyflag**
Dependent on: https://github.com/GrapheneOS/platform_packages_apps_Settings/pull/158
Needs testing!

### CODE REVIEW

File: core/java/android/provider/Settings.java [hunk 1]
```diff
@@ -17330,6 +17330,12 @@ public static boolean putFloat(ContentResolver cr, String name, float value) {
          */
         public static final String BLUETOOTH_OFF_TIMEOUT = "bluetooth_off_timeout";
 
+        /**
+         * The amount of time in milliseconds before the clipboard is cleared.
+         * @hide
+         */
+        public static final String CLIPBOARD_AUTO_CLEAR_TIMEOUT = "clipboard_auto_clear_timeout";
```
Comment (x86pup) [inline core/java/android/provider/Settings.java]: ExtSettings infrastructure does not touch Settings.java, you need to define this in [ExtSettings.java](https://github.com/GrapheneOS/platform_frameworks_base/blob/13/core/java/android/ext/settings/ExtSettings.java) instead

File: services/core/java/com/android/server/clipboard/ClipboardService.java [hunk 2]
```diff
@@ -429,9 +429,7 @@ private void scheduleAutoClear(@UserIdInt int userId, int intendingUid) {
         }
 
         private long getTimeoutForAutoClear() {
-            return DeviceConfig.getLong(DeviceConfig.NAMESPACE_CLIPBOARD,
-                    PROPERTY_AUTO_CLEAR_TIMEOUT,
-                    DEFAULT_CLIPBOARD_TIMEOUT_MILLIS);
+            return Settings.Global.getLong(mContext.getContentResolver(), Settings.Global.CLIPBOARD_AUTO_CLEAR_TIMEOUT, DEFAULT_CLIPBOARD_TIMEOUT_MILLIS);
```
Comment (x86pup) [inline services/core/java/com/android/server/clipboard/ClipboardService.java]: ExtSettings checks should import `import android.ext.settings.ExtSettings;` and do:

`ExtSettings.<key>.get(<context>)`

In this case:

`return ExtSettings.CLIPBOARD_AUTO_CLEAR_TIMEOUT.get(mContext)`

As we're not using Settings.java.

File: core/java/android/ext/settings/ExtSettings.java [hunk 3]
```diff
@@ -9,6 +9,7 @@
 
 /** @hide */
 public class ExtSettings {
+    public static final int DEFAULT_CLIPBOARD_TIMEOUT_MILLIS = 3600000;
```
Comment (x86pup) [inline core/java/android/ext/settings/ExtSettings.java]: ExtSettings constants should be defined in their own file, but in this case I don't think this needs to be a constant. In fact, we could just use the public constant from frameworks/base/services/core/java/com/android/server/clipboard/ClipboardService.java

`import static com.android.server.clipboard.ClipboardService.DEFAULT_CLIPBOARD_TIMEOUT_MILLIS`

File: core/java/android/ext/settings/ExtSettings.java [hunk 4]
```diff
@@ -31,6 +32,9 @@ public class ExtSettings {
     public static final BoolSetting SCRAMBLE_PIN_LAYOUT = new BoolSetting(
             Setting.Scope.PER_USER, "lockscreen_scramble_pin_layout", false);
 
+    public static final IntSetting CLIPBOARD_AUTO_CLEAR_TIMEOUT = new IntSetting(
+            Setting.Scope.GLOBAL, "clipboard_auto_clear_timeout", DEFAULT_CLIPBOARD_TIMEOUT_MILLIS);
```
Comment (x86pup) [inline core/java/android/ext/settings/ExtSettings.java]: I think we also want this to be per-user instead of global

**Review by x86pup — CHANGES_REQUESTED** [review]
Needs some changes, needs to be rebased, and I think we want this to be per-user instead of global unless due to some limitation.

I did type these suggestions on GitHub which may be incorrect / hard to understand, so if you want I can push my requested changes to my fork.

File: core/java/android/ext/settings/ExtSettings.java (hunk 3 above)
Comment (secrecyflag) [reply core/java/android/ext/settings/ExtSettings.java]: I tried that, but for some reason I couldn't figure out, `make` complains that `com.android.server.clipboard` doesn't exist?? I'm not sure if the package should be added to `Android.bp`, but I left it for now this way. Could you verify it?

Comment (x86pup) [reply core/java/android/ext/settings/ExtSettings.java]: Shouldn't need to touch Android.bp, but I can look at it later today

Comment (secrecyflag) [reply core/java/android/ext/settings/ExtSettings.java]: Solved by moving the constant to a new file.

**Comment by secrecyflag** [comment]
Should be ready for testing on a pixel.

**Comment by x86pup** [comment]
Still needs to be rebased

**Comment by x86pup** [comment]
Toggle doesn't seem to show up in my Pixel 6 after building.

(See review in Settings for solution)

File: services/core/java/com/android/server/clipboard/ClipboardService.java [hunk 5]
```diff
@@ -429,9 +430,7 @@ private void scheduleAutoClear(@UserIdInt int userId, int intendingUid) {
         }
 
         private long getTimeoutForAutoClear() {
```
Comment (x86pup) [inline services/core/java/com/android/server/clipboard/ClipboardService.java]: Seems like this isn't a per-user Context (and mContext seems invalid here, build error) which will cause the settings in the Owner profile to apply across all profiles irrespective of the current profile's setting, so you'll need to specify the uid this is for.

```suggestion
        private long getTimeoutForAutoClear(int userId) {
            return ExtSettings.CLIPBOARD_AUTO_CLEAR_TIMEOUT.get(getContext(), userId);
```

And where `getTimeoutForAutoClear` method is called, pass the `userId` argument:

`getTimeoutForAutoClear(userId)`

**Review by x86pup — CHANGES_REQUESTED** [review]
Need to replace `mContext` with `getContext()` method, and get the userId as the Context is not per-user

File: services/core/java/com/android/server/clipboard/ClipboardService.java (hunk 5 above)
Comment (secrecyflag) [reply services/core/java/com/android/server/clipboard/ClipboardService.java]: Didn't catch that build error as I'm currently using `m Settings`. Waiting for a new SSD to arrive (don't have enough space on device).

**Comment by secrecyflag** [comment]
Closing this pull request as it became messy. Will open a new one and send here.

**Comment by secrecyflag** [comment]
#357 @r3g-5z

**Pull request closed without merging**